### PR TITLE
fix: preserve trace if context if closed manually during test run

### PR DESCRIPTION
**Pull request by yury-s**
Reference #22122, #22120

### CODE REVIEW

File: packages/playwright-test/src/worker/workerMain.ts [hunk 1]
```diff
@@ -369,8 +369,13 @@ export class WorkerMain extends ProcessRunner {
       await testInfo._runFn(async () => {
         // Now run the test itself.
         debugTest(`test function started`);
+        setRunningTestBody(true);
```
Comment (dgozman) [inline packages/playwright-test/src/worker/workerMain.ts:372]: Let's assign a property on `TestInfoImpl` instead.

File: packages/playwright-test/src/index.ts [hunk 2]
```diff
@@ -331,7 +332,10 @@ const playwrightFixtures: Fixtures<TestFixtures, WorkerFixtures> = ({
       (tracing as any)[startedCollectingArtifacts] = true;
       if (captureTrace) {
         let tracePath;
-        if (preserveTrace()) {
+        // Create a trace file if we know that:
+        // - it is's going to be used due to the config setting and the test status or
+        // - we are inside a test and the user manually closed the context.
+        if (preserveTrace() || runningTestBody()) {
```
Comment (dgozman) [inline packages/playwright-test/src/index.ts:338]: I guess we can also have a new context created and closed in `beforeEach`/`afterEach` hook, and we should preserve its trace as well? Perhaps just check whether we have started the fixture teardown phase or not.

File: packages/playwright-test/src/index.ts [hunk 3]
```diff
@@ -24,6 +24,7 @@ import type { TestInfoImpl } from './worker/testInfo';
 import { rootTestType } from './common/testType';
 import { type ContextReuseMode } from './common/config';
 import { artifactsFolderName } from './isomorphic/folders';
+import { runningTestBody } from './common/globals';
```
Comment (pavelfeldman) [inline packages/playwright-test/src/index.ts:27]: This looks like a layering violation to me: index.ts is a part of Playwright, globals are a part of folio. Why do we need to know the stage of the test if we are the ones that start and end the tracing?

File: packages/playwright-test/src/worker/workerMain.ts (hunk 1 above)
Comment (yury-s) [reply packages/playwright-test/src/worker/workerMain.ts]: Added one on the context instead.

File: packages/playwright-test/src/index.ts (hunk 3 above)
Comment (yury-s) [reply packages/playwright-test/src/index.ts]: We don't.

**Review by dgozman — APPROVED** [review]

**Pull request merged**